File: modules/TechSpecGenerator/src/types/workflow-state.ts
/**
 * TypeScript equivalent of Python workflow state management
 */

export interface CrawlResult {
  success: boolean;
  filepath?: string;
  content_length: number;
  error?: string;
  url: string;
}

export interface WorkflowMetadata {
  start_time?: number;
  end_time?: number;
  duration?: number;
  total_urls?: number;
  successful_crawls?: number;
  failed_crawls?: number;
  spec_generated?: boolean;
  estimated_tokens?: Record<string, number>;
  mock_server_generated?: boolean;
}

export interface NodeConfig {
  enabled: boolean;
  retry_count?: number;
  timeout?: number;
  parallel?: boolean;
  custom_params?: Record<string, any>;
}

export interface WorkflowConfig {
  url_collection?: NodeConfig;
  crawling?: NodeConfig;
  llm_processing?: NodeConfig;
  mock_server_generation?: NodeConfig;
  output_management?: NodeConfig;
}

export interface Config {
  firecrawl: {
    api_key: string;
  };
  litellm: {
    model: string;
    api_key: string;
    max_tokens: number;
    base_url?: string;
    custom_headers?: Record<string, string>;
  };
  workflow: WorkflowConfig;
}

export interface MockServerData {
  server_js: string;
  package_json: string;
  info: string;
}

export interface WorkflowState {
  // Configuration
  config: Config;
  output_dir: string;
  
  // Input data
  urls: string[];
  
  // Processing results
  crawl_results: Record<string, CrawlResult>;
  markdown_files: string[];
  tech_spec?: string;
  spec_filepath?: string;
  
  // Mock server results
  mock_server_dir?: string;
  mock_server_process?: any;
  mock_server_data?: MockServerData;
  
  // Error tracking
  errors: string[];
  warnings: string[];
  
  // Workflow metadata
  metadata: WorkflowMetadata;
  
  // Node control flags
  node_config?: Record<string, Record<string, any>>;
}

export type NodeFunction = (state: WorkflowState) => Promise<WorkflowState>;
export type ConditionalFunction = (state: WorkflowState) => string;

File: modules/TechSpecGenerator/src/api_doc_processor/workflow_state.py
"""State management for the LangGraph workflow."""

from pathlib import Path
from typing import Dict, List, Any, Optional
from typing_extensions import TypedDict, NotRequired

from .config import Config


class CrawlResult(TypedDict):
    """Result from crawling a single URL."""
    success: bool
    filepath: Optional[str]
    content_length: int
    error: Optional[str]
    url: str


class WorkflowMetadata(TypedDict):
    """Metadata about the workflow execution."""
    start_time: NotRequired[float]
    end_time: NotRequired[float]
    duration: NotRequired[float]
    total_urls: NotRequired[int]
    successful_crawls: NotRequired[int]
    failed_crawls: NotRequired[int]
    spec_generated: NotRequired[bool]
    estimated_tokens: NotRequired[Dict[str, int]]
    mock_server_generated: NotRequired[bool]


class WorkflowState(TypedDict):
    """Complete state for the API documentation processing workflow."""
    
    # Configuration
    config: Config
    output_dir: Path
    
    # Input data
    urls: List[str]
    
    # Processing results
    crawl_results: Dict[str, CrawlResult]
    markdown_files: List[Path]
    tech_spec: NotRequired[str]
    spec_filepath: NotRequired[Path]
    
    # Mock server results
    mock_server_dir: NotRequired[Path]
    mock_server_process: NotRequired[Any]
    mock_server_data: NotRequired[Dict[str, Any]]
    
    # Error tracking
    errors: List[str]
    warnings: List[str]
    
    # Workflow metadata
    metadata: WorkflowMetadata
    
    # Node control flags
    node_config: NotRequired[Dict[str, Dict[str, Any]]]


class NodeConfig(TypedDict):
    """Configuration for individual workflow nodes."""
    enabled: bool
    retry_count: NotRequired[int]
    timeout: NotRequired[float]
    parallel: NotRequired[bool]
    custom_params: NotRequired[Dict[str, Any]]


class WorkflowConfig(TypedDict):
    """Configuration for the entire workflow."""
    url_collection: NotRequired[NodeConfig]
    crawling: NotRequired[NodeConfig] 
    llm_processing: NotRequired[NodeConfig]
    mock_server_generation: NotRequired[NodeConfig]
    output_management: NotRequired[NodeConfig]